Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -76,6 +110,112 @@ TextureMapperLayer::~TextureMapperLayer()
     removeFromParent();
 }
 
+void TextureMapperLayer::preprocess(TextureMapperPaintOptions& options)
+{
+    // Preprocess is called as a first step before painting any layers
+
+    // Layers are collected in depth first, post-order
+    Vector<TextureMapperLayer*> layersToFlatten;
+    collectLayersToFlattenRecursive(layersToFlatten);
+
+    for (auto* layer : layersToFlatten)
+        layer->flatten(options);
+}
+
+void TextureMapperLayer::collectLayersToFlattenRecursive(Vector<TextureMapperLayer*>& layers)
+{
+    // Traverse all children
+    for (auto* child : m_children) {
+        // Recursively collect from the child's descendants
+        child->collectLayersToFlattenRecursive(layers);
+
+        // If the child satisfies the condition, add it to the layers vector
+        if (child->flattensAsLeafOf3DSceneOr3DPerspective())
+            layers.append(child);
+    }
+}
+
+void TextureMapperLayer::flatten(TextureMapperPaintOptions& options)
+{
+    // Reset transformations for flattening root
+    SetForScope scopedLocalTransform(m_layerTransforms.localTransform, TransformationMatrix());
+    SetForScope scopedFutureLocalTransform(m_layerTransforms.futureLocalTransform, TransformationMatrix());
+    m_layerTransforms.combined = TransformationMatrix();
+    m_layerTransforms.combinedForChildren = TransformationMatrix();
+    m_layerTransforms.futureLocalTransform = TransformationMatrix();
+    m_layerTransforms.futureCombined = TransformationMatrix();
+    m_layerTransforms.futureCombinedForChildren = TransformationMatrix();
+    ComputeTransformData data;
+    if (m_state.maskLayer)
+        m_state.maskLayer->computeTransformsRecursive(data);
+    if (m_state.replicaLayer)
+        m_state.replicaLayer->computeTransformsRecursive(data);
+    if (m_state.backdropLayer)
+        m_state.backdropLayer->computeTransformsRecursive(data);
+    for (auto* child : m_children)
+        child->computeTransformsRecursive(data);
+
+    // Flatten the layer
+    if (m_flattenedLayer != nullptr) {
+        delete m_flattenedLayer;
```

Comment:
No manual memory management in WebKit code.

Reply:
Fixed